sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,841 @@
+composite specialization constants do not have separate ID and can only be set
+by setting value to each member of a composite, which means that we have `1:n`
+mapping between SYCL identifiers and numeric IDs of specialization constants.
+
+4. When AOT compilation is used or target is a CUDA device (where NVPTX
+intermediate representation is used), we need to somehow emulate support for
+specialization constants.
+
+## Design
+
+As stated above, native specialization constants support is based on
+corresponding SPIR-V functionality, while emulation is supposed to be
+implemented through transforming specialization constants into kernel arguments.
+
+In DPC++ Headers/DPC++ RT we don't know a lot of necessary information about
+specialization constants, like: which numeric ID is used for particular
+specialization constant (since we support `SYCL_EXTERNAL`, those IDs can only
+be allocated by the compiler during link stage) or which kernel argument is used
+to pass particular specialization constant (because they are not explicitly
+captured by SYCL kernel functions and regular mechanism for kernel arguments
+handling can't be used here).
+
+Therefore, we can't have headers-only implementation and the crucial part of
+design is how to organize mapping mechanism between SYCL identifiers for
+specialization constants (`specialization_id`s) and low-level identifiers
+(numeric IDs in SPIR-V or information about corresponding kernel arguments).
+
+That mapping mechanism is particularly tricky, because of some additional
+complexity coming from SYCL 2020 specification:
+- `specialization_id` variables, which are used as specialization constant
+  identifiers (being non-type template parameters of some methods) can't be
+  forward-declared in general case (for example, if defined as `static`), which
+  means that we can't use integration header to attach some information to them
+  through some C++ templates tricks (like it is done for regular kernel
+  arguments or kernel names, for example).
+- they also can be declared as `static` or just non-`inline` `constexpr`, which
+  means that they have internal linkage and can't be referenced from other
+  translation units, which means that we can't for example create a new
+  translation unit which contains some mapping from `specialization_id` address
+  to some desired info.
+
+Based on those limitations, the following mapping design is proposed:
+- DPC++ RT uses special function:
+  ```
+  namespace detail {
+    template<auto &SpecName>
+    const char *get_spec_constant_symbolic_ID();
+  }
+  ```
+  Which is only declared, but not defined in there and used to retrieve required
+  information like numeric ID of a specialization constant.
+- Definition of that function template are provided by DPC++ FE in form of
+  _integration footer_: the compiler generates a piece of C++ code which is
+  injected at the end of a translation unit:
+  ```
+  namespace detail {
```

Comment:
This section is a little odd... where did the 'float' version come from?

Reply:
The comment below says that we assume that user has two `specialization_id` definitions in the code:
```
constexpr specialiation_id<int> int_const;
class Wrapper {
public:
 static constexpr specialization_id<float> float_const;
};
```

`float` specialization constant comes from that `Wrapper` class

Reply:
Thanks, I misread it thanks to mis-interpreting comment markers in a couple of places.